### PR TITLE
Implement Copy Elimination for dpdk backend

### PR DESCRIPTION
**Pull request by kamleshbhalui**
consider only moves and cast statements.

algorithm:
1) collect all used, copied and defined variables in all statements in a block
2)iterate over list of statements in a block in reverse order
3)if a statement is mov/cast, check if destination variable is a single definition, has single use, and also is a single copy
then drop that statement and update all use point of destintation with source of mov/cast operation.
4) else copy statement and push to the to the list.
5) create the list in original order and return.

### CODE REVIEW

File: backends/dpdk/dpdkAsmOpt.h [hunk 1]
```diff
@@ -290,6 +290,58 @@ class ShortenTokenLength : public Transform {
     }
 };
 
+// This pass identify redundant copies/moves and eliminate it
+// and propogate original sources.
+class CopyPropagationAndElimination : public Transform {
+    std::unordered_map<cstring, int> copyInfo;
```
Comment (kamleshbhalui) [inline backends/dpdk/dpdkAsmOpt.h:296]: store destination operand of mov and cast ops.

File: backends/dpdk/dpdkAsmOpt.h [hunk 2]
```diff
@@ -290,6 +290,58 @@ class ShortenTokenLength : public Transform {
...
+    std::unordered_map<cstring, int> usesInfo;
```
Comment (kamleshbhalui) [inline backends/dpdk/dpdkAsmOpt.h:297]: store source operand of all binary, unary , cond jumps

File: backends/dpdk/dpdkAsmOpt.h [hunk 3]
```diff
@@ -290,6 +290,58 @@ class ShortenTokenLength : public Transform {
...
+    std::unordered_map<cstring, int> defInfo;
```
Comment (kamleshbhalui) [inline backends/dpdk/dpdkAsmOpt.h:298]: store destination operand of all ops

File: backends/dpdk/dpdkAsmOpt.h [hunk 4]
```diff
@@ -290,6 +290,58 @@ class ShortenTokenLength : public Transform {
...
+    std::unordered_map<cstring /*def*/, const IR::Expression* /*use*/> replacementMap;
```
Comment (kamleshbhalui) [inline backends/dpdk/dpdkAsmOpt.h:299]: maps for replacing a use with it's def.

File: backends/dpdk/dpdkAsmOpt.h [hunk 5]
```diff
@@ -290,6 +290,58 @@ class ShortenTokenLength : public Transform {
...
+    void markUseDef(const IR::DpdkMovStatement *mv);
+    void markUseDef(const IR::DpdkUnaryStatement *u);
+    void markUseDef(const IR::DpdkCastStatement *c);
+    void markUseDef(const IR::DpdkJmpCondStatement *b);
+    const IR::Expression* replaceIfCopy(const IR::Expression *expr);
+
+    IR::IndexedVector<IR::DpdkAsmStatement>
+    copyPropAndDeadCodeElim(IR::IndexedVector<IR::DpdkAsmStatement> stmts);
+    void clear() {
+        replacementMap.clear();
```
Comment (kamleshbhalui) [inline backends/dpdk/dpdkAsmOpt.h:317]: clear all data structure as we consider every action and apply block a independent block.

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 6]
```diff
@@ -264,5 +264,120 @@ bool ValidateTableKeys::preorder(const IR::DpdkAsmProgram *p) {
...
 }
 
+const IR::Expression* CopyPropagationAndElimination::getIrreplaceableExpr(cstring str) {
+    if (replacementMap.count(str) == 0)
+        return nullptr;
+    if (!str.startsWith("m."))
+        return nullptr;
+    auto expr = replacementMap[str];
+    const IR::Expression* prev = nullptr;
+    while (expr != nullptr && expr->is<IR::Member>() && str != expr->toString()
```
Comment (kamleshbhalui) [inline backends/dpdk/dpdkAsmOpt.cpp:274]: try to find expression which will replace `str` ,

**Review by usha1830 — APPROVED** [review]
Looks good. Have some minor comments. Please check

File: testdata/p4_16_samples_outputs/pna-add-on-miss.p4.spec [hunk 7]
```diff
@@ -60,9 +57,8 @@ action next_hop2 args instanceof next_hop2_arg_t {
 }
 
 action add_on_miss_action2 args none {
-	mov m.MainControlT_tmp 0x0
 	mov m.MainControlT_tmp_0 0x4d2
-	learn next_hop2 m.MainControlT_tmp m.local_metadata_timeout
+	learn next_hop2 0x0 m.local_metadata_timeout
```
Comment (usha1830) [inline testdata/p4_16_samples_outputs/pna-add-on-miss.p4.spec:61]: This is incorrect. Learn instruction doesn't accept immediate

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 8]
```diff
@@ -264,5 +264,120 @@ bool ValidateTableKeys::preorder(const IR::DpdkAsmProgram *p) {
...
+        } else if (auto c = stmt->to<IR::DpdkCastStatement>()) {
+            markUseDef(c);
+        } else if (auto jc = stmt->to<IR::DpdkJmpCondStatement>()) {
+            markUseDef(jc);
+        }
+    }
+    IR::IndexedVector<IR::DpdkAsmStatement> instr;
+    for (auto stmt1 = stmts.rbegin(); stmt1!=stmts.rend();stmt1++) {
+        auto stmt = *stmt1;
+        if (auto mv = stmt->to<IR::DpdkMovStatement>()) {
```
[the remark this answers is not in the document]
Comment (kamleshbhalui) [reply backends/dpdk/dpdkAsmOpt.cpp]: done

File: testdata/p4_16_samples_outputs/pna-example-varIndex-1.p4.spec [hunk 9]
```diff
@@ -99,8 +96,7 @@ action execute_1 args none {
 	mov m.MainControlT_tmp_16 h.vlan_tag_0.pcp_cfi_vid
 	shr m.MainControlT_tmp_16 0x4
 	mov m.MainControlT_tmp_17 m.MainControlT_tmp_16
```
[the remark this answers is not in the document]
Comment (kamleshbhalui) [reply testdata/p4_16_samples_outputs/pna-example-varIndex-1.p4.spec]: done

File: testdata/p4_16_samples_outputs/pna-add-on-miss.p4.spec (hunk 7 above)
Comment (kamleshbhalui) [reply testdata/p4_16_samples_outputs/pna-add-on-miss.p4.spec:61]: ok

Comment (kamleshbhalui) [reply testdata/p4_16_samples_outputs/pna-add-on-miss.p4.spec]: done

**Pull request closed without merging**